pyccel/ast/numpyext.py

```diff
@@ -328,6 +330,21 @@ class NumpyComplex128(NumpyComplex):
 #==============================================================================
 
 def process_dtype(dtype):
+    """
+    Returns the corresponding type and precision to the dtype.
+
+    Args:
+        dtype     : Datatype
+
+    Raises:
+        TypeError: In the case of unrecognized argument type.
+        TypeError: In the case of passed string argument not recognized as valid dtype.
+
+    Returns:
+        dtype     : Datatype
+        precision : int
+    """
```

Comment:
We try to use the same formatting for all our docstrings which allows them to be recognised by doxygen if we ever have enough documentation to make it worth adding automated documentation. See [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/core.py#L164-L188) for an example of a docstring.

After reading the docstring we should understand:
- [x] what the purpose of the function is
- [ ] In what context the function should be used
- [ ] What the expected type of each of the arguments are
- [ ] What each of the arguments is for
- [x] What the expected type of each of the results are

Currently we haven't added a section for possible errors but I think this could be useful so I would keep this.

I think the information missing from your docstring is:
- Argument type (If `dtype` is a `DataType` then a `TypeError` is raised so the type that you have written is not the expected type)
- Argument role
- The description could be expanded to explain where this function should be used (i.e. to handle `dtype` arguments of array creation functions). The explanation should be clear enough to avoid the confusion that you had when you started tackling this issue.

Reply:
```
    """
    This function will take a dtype passed to a numpy array creation function
    and depending on the type of the passed argument, it will do the necessary treatments
    to extract the corresponding type and precision from it then return the result.

    This function could be useful when working with numpy creation function having a dtype argument,
    like numpy.array, nump.arrange, numpy.linespace...
    
    Parameters:
    ----------
        dtype     : PythonType | PyccelFunctionDef | String | Numpy Datatype | Python Builtin Datatype
        The actual dtype passed to the numpy function

    Raises:
    ----------
        TypeError : In the case of unrecognized argument type.
        TypeError : In the case of passed string argument not recognized as valid dtype.

    Returns:
    ----------
        dtype     : Datatype
        The Datatype corresponding to the passed dtype.
        precision : int
        The precision corresponding to the passed dtype.
    """
```
Is something like this better ?

Reply:
That looks a lot better :)

Typo `linespace` -> `linspace`

What `PyccelFunctionDef` other than a `Numpy Datatype` or a `Python Builtin Datatype` can be used as an argument?

Reply:
From what i know it can only be one of those two.

Reply:
Then I don't think you need to mention `PyccelFunctionDef` separately.
You should have one of:
```
dtype     : PythonType | PyccelFunctionDef | String
```
```
dtype     : PythonType | String | Numpy Datatype | Python Builtin Datatype
```
```
dtype     : PythonType | PyccelFunctionDef (Numpy Datatype or Python Builtin Datatype) | String
```